src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutorKeepalived.java

```diff
@@ -0,0 +1,272 @@
+        throws IOException {
+      try {
+        while (operations.hasNext()) {
+          operation = operations.next();
+
+          if (resetBackoff) {
+            // Assuming the server has made progress since we received the response. Reset the backoff
+            // so that this request has a full deck of retries
+            backoff.reset();
+          }
+
+          ExecuteResponse response = handleOperation(operation, observer);
+          if (response != null) {
+            return response;
+          }
+        }
+
+        // The operation completed successfully but without a result.
+        throw new IOException("Remote server error: execution terminated with no result.");
+      } finally {
+        close(operations);
+      }
+    }
+
+    void close(Iterator<Operation> operations) {
+      // The blocking streaming call closes correctly only when trailers and a Status are received
+      // from the server so that onClose() is called on this call's CallListener. Under normal
+      // circumstances (no cancel/errors), these are guaranteed to be sent by the server only if
+      // operations.hasNext() has been called after all replies from the stream have been
+      // consumed.
+      try {
+        while (operations.hasNext()) {
+          operations.next();
+        }
+      } catch (StatusRuntimeException e) {
+        // Cleanup: ignore exceptions, because the meaningful errors have already been propagated.
+      }
+    }
+
+    static void throwIfError(Status status, @Nullable ExecuteResponse resp) {
+      if (status.getCode() == Code.OK.value()) {
+        return;
+      }
+      throw new ExecutionStatusException(status, resp);
+    }
+
+    @Nullable
+    static ExecuteResponse handleOperation(Operation operation, OperationObserver observer)
+        throws IOException {
+      // Update execution progress to the caller.
+      //
+      // After called `execute` above, the action is actually waiting for an available gRPC
+      // connection to be sent. Once we get a reply from server, we know the connection is up and
+      // indicate to the caller the fact by forwarding the `operation`.
+      //
+      // The accurate execution status of the action relies on the server
+      // implementation:
+      //   1. Server can reply the accurate status in `operation.metadata.stage`;
+      //   2. Server may send a reply without metadata. In this case, we assume the action is
+      //      accepted by the server and will be executed ASAP;
+      //   3. Server may execute the action silently and send a reply once it is done.
+      observer.onNext(operation);
+
+      if (operation.getResultCase() == Operation.ResultCase.ERROR) {
+        throwIfError(operation.getError(), null);
+      }
+
+      if (operation.getDone()) {
+        Preconditions.checkState(operation.getResultCase() != Operation.ResultCase.RESULT_NOT_SET);
```

Comment:
If this precondition check fails, it causes Bazel to crash. Looks like this come from the server, which would mean that the server can give inputs to Bazel that cause it to crash?

Reply:
Good catch! Changed to throw ExecutionStatusException (which will be wrapped in an IOException).